File: Solutions/31. Next Permutation/CPP/Next_Permutation.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

class Solution{
public:
    void nextPermutation(vector<int> &nums){
        vector<int>::iterator i = nums.end() - 1;

        while(i != nums.begin() and *(i - 1) >= *i){
            i = i - 1;
        }

        reverse(i, nums.end());
        
        if(i != nums.begin()){
            iter_swap(i - 1, upper_bound(i, nums.end(), *(i - 1)));
        }
    }
};

int main(){
    Solution *S = new Solution();

    //input
    vector<int> nums{1, 4, 5, 3, 2};

    S->nextPermutation(nums);

    for(auto i : nums){
        cout << i << ' ';
    }
    cout << endl;

    return 0;
}

File: Solutions/31. Next Permutation/Python/Next_Permutation.py
class Solution:
    def nextPermutation(self, nums: list[int]) -> None:
        nums_size = len(nums)
        for i in range(nums_size - 1, 0, -1):
            if nums[i - 1] < nums[i]:
                nums[i:] = sorted(nums[i:])
                for j in range(i, nums_size):
                    if nums[j] > nums[i - 1]:
                        nums[j], nums[i - 1] = nums[i - 1], nums[j]
                        return nums

        return nums.reverse()

S = Solution()

#inputs
nums = [3, 2, 1]

S.nextPermutation(nums)

print(nums)